### PR TITLE
Optimise TextEdit theme loading

### PR DESCRIPTION
**Pull request by Paulb23**
During editor load and when switching between scripts tabs, `TextEdit` was updating its cache for no reason, this was quite slow.

This PR adds two new methods to `Control`. `begin_bulk_theme_override`, which will prevent `*_theme_*_override` methods from causing a theme update, until `end_bulk_theme_override` is called.

Secondly, since we now have a dedicated code editing node `CodeEdit`, unlike before with `TextEdit`, we had to override most items to turn it into the code editor. Now there is no longer a need to do so, the theme overrides have been moved into the default editor theme.

Finally, `TextEdits` internal `text` class, now has a `is_dirty` flag to prevent invalidating the cache when not needed.

Overall, this has increased performance for a large script on load from `450ms` to `30ms`, and switching script tabs, from `1.3 seconds`  to `80ms`.

Seems to alleviate some of the issues seen in #51349

### CODE REVIEW

**Comment by AnilBK** [comment]
Your second commit "Only update TextEdit text cache when dirty." seems interesting.
Some Setters of nodes in scene/ folder, just set the property and update/regenerate the internal values again, without checking if the newly set values is same as old values.

I wonder if  we apply  condition in setters to check if new value is same as old values wherever possible, how much performance gains would be available.

**Comment by akien-mga** [comment]
See #49227, it seems you're reimplementing something which already exists for preventing theme updates (CC @pycbouh).

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -105,7 +129,11 @@ int TextEdit::Text::get_line_height(int p_line, int p_wrap_index) const {
 }
 
 void TextEdit::Text::set_width(float p_width) {
+	if (width == p_width) {
+		return;
+	}
 	width = p_width;
+	is_dirty = true;
```
Comment (bruvzg) [inline scene/gui/text_edit.cpp:136]: `set_width` and `set_tab_size` do not fully invalidate cache, it's calling `invalidate_all_lines()` instead (which resetting line breaks only).

**Comment by pycbouh** [comment]
> See #49227, it seems you're reimplementing something which already exists for preventing theme updates (CC @pycbouh).

Similar, but not quite the same. What I've added in the linked PR prevents changes to the `Theme` resource from propagating during bulk operations. This one solves the situation where local overrides on a control are updated and cause the theme changed signal to fire, unrelated to the `Theme` resource.

Now, the only reason why the control fires that notification at all is so it can update itself (the signal is not propagated to the children, because they don't really care for the local overrides on the parent control). So this solution seems fine and not stepping on the same ground as the linked PR. Though I'm not sure if we should expose the new methods to the scripting API, as they need to be handled with care. Or at the very least a big warning is in order in the documentation.

**Review by bruvzg — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by Chaosus** [comment]
Oh, seems like this fixes the https://github.com/godotengine/godot/issues/51349 - it speeds up it very much at least!

**Pull request merged**